src/_gettsim/shared.py

```diff
@@ -24,12 +24,16 @@ def add_rounding_spec(params_key):
     Returns
     -------
     func : function
-        Function with __rounding_params_key__ attribute
+        Function with __gettsim__["rounding_params_key"] attribute
 
     """
 
     def inner(func):
```

Comment:
So IIUC, there could be an `@wraps(func)` just before this function definition (possibly with appropriate arguments) and we would not need the code below the line `# Remember data from decorator` anymore?

@janosg @timmens, this might be interesting for you guys, too!

Reply:
Although the example from the [docs](https://docs.python.org/3/library/functools.html#functools.wraps) suggests that using `@wraps` will require (hopefully only) an extra line for the `change_name` keyword of `@dates_active` (#495)?

Reply:
> So IIUC, there could be an `@wraps(func)` just before this function definition (possibly with appropriate arguments) and we would not need the code below the line `# Remember data from decorator` anymore?

The wraps doesn't help in this case since 
1. we are potentially creating the `dict` here the first time or adding a new entry,
2. we return `func` itself instead of a new function that calls `func` internally.

Reply:
@hmgaudecker are you suggesting to have our own wraps function that creates the `__info__` dict if not present and copies it otherwise?

Reply:
I actually did not fully see through it, but I did think something like this would be helpful. I.e., ideally we'll want to make sure that every function in the DAG has an `__info__` dict so we can just call `func.__info__.get(key)` without additional checks. 

This should also happen for functions the user puts into the DAG.

More of a note, probably for a different PR?

Reply:
While we evaluate the decorators we need the `hasattr` check since the decorators can be applied in any order, so `__info__` might not exist yet. I also don't see a good way of storing their data in a place other than a field of the decorated functions.

Afterwards, however, I absolutely agree that ensuring
1. that the dict `__info__` exists,
2. that it has specific keys like `rounding_params_key`
before accessing the values gets annoying. 

I also don't like storing data with a fixed structure in dictionaries since
1. accessing it using strings is brittle (linters don't catch typos),
2. IDEs don't offer robust auto-completion or refactorings,
3. any behavior that uses the data must be defined externally, so the structure of the data becomes part of the interface.

I was thinking about creating a class `PolicyFunction` that contains the actual callable and any additional data. The `function_loader` could then collect the functions (their decorators are applied automatically) and then create `PolicyFunction` objects. Afterwards, the code could work with the created objects.

Could be something along the lines of:

```py
class PolicyFunction:
    def __init__(
        self,
        f: Callable,
        dag_key: str,
        start_date: datetime.date,
        end_date: datetime.date
    ):
        ...

    def is_active_at_date(date: datetime.date) -> bool:
        ... 
```

Reply:
But yes, that would be another PR. I could create an issue for it.

Reply:
> But yes, that would be another PR. I could create an issue for it.

Sounds good. We should think about whether to have a base class in dags then, subclassing it here.

Reply:
> I also don't like storing data with a fixed structure in dictionaries since
> 
>     1. accessing it using strings is brittle (linters don't catch typos),
> 
>     2. IDEs don't offer robust auto-completion or refactorings,
> 
>     3. any behavior that uses the data must be defined externally, so the structure of the data becomes part of the interface.
> 
> ...
I understand your points. I definitely agree with the first two, but not necessarily with the last one. I think a well designed public data interface might be just the thing we want here!

But I think the idea here is that the `__info__` dict does not have a fixed structure. Any decorator in `gettsim` or `dags` might add something to it. But (by convention) None would delete or overwrite anything that is there.

Reply:
I'll tend to side with Lars here, though I shall be happy to be convinced otherwise.

In general, well-defined interfaces are a good thing. 

If we add a decorator in GETTSIM that needs a new field, we would add that to the class. 

If dags adds an attribute we don't care about, no problem, just ignore (if we already have that, it will be much more obvious we would be overwriting it).

If users want to do magic like this, they'll better do it in their own data structures than to hijack the dict.

Am I missing something?